server/src/main/java/org/elasticsearch/action/bulk/Retry2.java

```diff
@@ -0,0 +1,303 @@
+     * either all requests have succeeded or all retry attempts have been exhausted.
+     */
+    private final class RetryHandler implements ActionListener<BulkResponse> {
+        private static final RestStatus RETRY_STATUS = RestStatus.TOO_MANY_REQUESTS;
+        private final BulkRequest bulkRequest;
+        private final BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer;
+        private final ActionListener<BulkResponse> listener;
+        private final List<BulkItemResponse> responsesAccumulator;
+        private final long startTimestampNanos;
+        private final int retriesRemaining;
+
+        /**
+         * Creates a RetryHandler listener
+         * @param bulkRequest The BulkRequest to be sent, a subset of the original BulkRequest.
+         * @param responsesAccumulator The accumulator of all BulkItemResponses for the original BulkRequest
+         * @param consumer
+         * @param listener The delegate listener
+         * @param retriesRemaining The number of retry attempts remaining for the bulkRequestForRetry
+         */
+        RetryHandler(
+            BulkRequest bulkRequest,
+            List<BulkItemResponse> responsesAccumulator,
+            BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+            ActionListener<BulkResponse> listener,
+            int retriesRemaining
+        ) {
+            this.bulkRequest = bulkRequest;
+            this.responsesAccumulator = responsesAccumulator;
+            this.consumer = consumer;
+            this.listener = listener;
+            this.startTimestampNanos = System.nanoTime();
+            this.retriesRemaining = retriesRemaining;
+        }
+
+        @Override
+        public void onResponse(BulkResponse bulkItemResponses) {
+            if (bulkItemResponses.hasFailures() == false) {
+                totalBytesInFlight.addAndGet(-1 * bulkRequest.estimatedSizeInBytes());
+                logger.trace(
+                    "Got a response in {} with {} items, no failures",
+                    bulkItemResponses.getTook(),
+                    bulkItemResponses.getItems().length
+                );
+                // we're done here, include all responses
+                addResponses(bulkItemResponses, (r -> true));
+                listener.onResponse(getAccumulatedResponse());
+            } else {
+                if (canRetry(bulkItemResponses)) {
+                    logger.trace(
+                        "Got a response in {} with {} items including failures, can retry",
+                        bulkItemResponses.getTook(),
+                        bulkItemResponses.getItems().length
+                    );
+                    addResponses(bulkItemResponses, (r -> r.isFailed() == false));
+                    BulkRequest retryRequest = createBulkRequestForRetry(bulkItemResponses);
+                    totalBytesInFlight.addAndGet(-1 * (bulkRequest.estimatedSizeInBytes() - retryRequest.estimatedSizeInBytes()));
```

Comment:
I don't think we need to be so fine-grained here. We can just track the whole (original) request until it's done.